### PR TITLE
Added more tests and fix issues

### PR DESCRIPTION
**Pull request by matriv**

### CODE REVIEW

**Comment by matriv** [comment]
@mfussenegger Please review

File: sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java [hunk 1]
```diff
@@ -231,7 +231,7 @@ protected static String getQuotedSubscriptLiteral(String nodeName) {
         }
     }
 
-    private class InnerExpressionAnalyzer extends AstVisitor<Symbol, ExpressionAnalysisContext> {
+    protected class InnerExpressionAnalyzer extends AstVisitor<Symbol, ExpressionAnalysisContext> {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java]: why is this necessary?

File: sql/src/main/java/io/crate/analyze/relations/RelationAnalyzer.java [hunk 2]
```diff
@@ -546,4 +547,28 @@ protected AnalyzedRelation visitTableSubquery(TableSubquery node, StatementAnaly
...
+                                                  convertParamFunction,
+                                              FieldProvider fieldProvider,
+                                              @Nullable SubqueryAnalyzer subQueryAnalyzer) {
+            super(functions, sessionContext, convertParamFunction, fieldProvider, subQueryAnalyzer);
+        }
+
+        @Override
+        protected Symbol convertFunctionCall(FunctionCall node, ExpressionAnalysisContext context) {
+            try {
+                // Try to process function call as a string based on the outputs of the 1st relation of the union
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/relations/RelationAnalyzer.java]: please add a comment explaining why this is necessary

File: sql/src/main/java/io/crate/analyze/relations/RelationNormalizer.java [hunk 3]
```diff
@@ -486,4 +487,32 @@ public AnalyzedRelation visitTwoRelationsUnion(TwoRelationsUnion twoTableUnion,
...
+            if (newFunction == null) {
+                return super.visitFunction(symbol, context);
+            }
+            return newFunction;
+        }
+
+        @Override
+        public Symbol visitSymbol(Symbol symbol, List<Symbol> context) {
+            Symbol newSymbol = InputColumn.fromSymbol(symbol, context);
+            if (newSymbol == null) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/relations/RelationNormalizer.java]: why isn't this an error case?

File: sql/src/main/java/io/crate/analyze/symbol/InputColumn.java [hunk 4]
```diff
@@ -75,7 +75,11 @@ public static void shiftRight(@Nullable Collection<Symbol> symbols) {
      * generate an inputColumn which points to some symbol that is part of sourceList
      */
     public static InputColumn fromSymbol(Symbol symbol, List<? extends Symbol> sourceList) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/symbol/InputColumn.java]: add nullable annotation and extend docs please

File: sql/src/main/java/io/crate/planner/UnionFlatteningVisitor.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+import io.crate.analyze.relations.QueriedDocTable;
+import io.crate.analyze.relations.QueriedRelation;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Processes a tree of Unions and returns a flat List of all sub-relations
+ */
+public class UnionFlatteningVisitor
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/planner/UnionFlatteningVisitor.java]: please hide the visitor implementation if not relevant for the API

Comment (matriv) [inline sql/src/main/java/io/crate/planner/UnionFlatteningVisitor.java]: I extracted it as public because it's needed in the UnionConsumer and SelectStatementPlanner

File: sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java (hunk 1 above)
Comment (matriv) [inline sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java]: Because otherwise I cannot use it here:
https://github.com/crate/crate/pull/4421/files/def177ff6d4857d1a9c2d1d9c37859717bb5c57f#diff-bf12c5e61f9b0fe950aa518b1ed601abR566

File: sql/src/main/java/io/crate/analyze/relations/RelationAnalyzer.java [hunk 6]
```diff
@@ -141,13 +138,21 @@ protected AnalyzedRelation visitQuery(Query node, StatementAnalysisContext state
                     node.getOffset().get(),
                     relationAnalysisContext.expressionAnalysisContext())));
             }
-            querySpec.orderBy(analyzeOrderBy(
+
+            ExpressionAnalyzer orderByExpressionAnalyzer = new ExpressionAnalyzer(
+                functions,
+                statementContext.sessionContext(),
+                statementContext.convertParamFunction(),
+                new UnionOrderByFieldProvider(relationAnalysisContext.sources()),
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/relations/RelationAnalyzer.java]: wouldn't it be enought to just create a new map of sources and use that with the existing FullQualifiedNameFieldProvider ?
